cadquery/cq.py

```diff
@@ -3441,48 +3478,133 @@ def loft(
+                ws[0].Center(), eDir, direction=direction
+            )
+            if len(facesList) == 0 and both:
+                raise ValueError(
+                    "Couldn't find a face to extrude/cut to for at least one of the two required directions of extrusion/cut."
+                )
+
+            if len(facesList) == 0:
+                # if we don't find faces in the workplane normal direction we try the other direction (as the user might have created a workplane with wrong orientation)
+                facesList = self.findSolid().facesIntersectedByLine(
+                    ws[0].Center(), eDir.multiply(-1.0), direction=direction
+                )
+                if len(facesList) == 0:
+                    raise ValueError(
+                        "Couldn't find a face to extrude/cut to. Check your workplane orientation."
+                    )
+            return facesList
+
         # group wires together into faces based on which ones are inside the others
         # result is a list of lists
 
         wireSets = sortWiresByBuildOrder(self.ctx.popPendingWires())
 
         # compute extrusion vector and extrude
-        eDir = self.plane.zDir.multiply(distance)
+        if upToFace is not None:
+            eDir = self.plane.zDir
+        elif distance is not None:
+            eDir = self.plane.zDir.multiply(distance)
+
+        if additive:
+            direction = "AlongAxis"
+        else:
+            direction = "Opposite"
 
         # one would think that fusing faces into a compound and then extruding would work,
-        # but it doesn't-- the resulting compound appears to look right, ( right number of faces, etc)
+        # but it doesnt-- the resulting compound appears to look right, ( right number of faces, etc)
         # but then cutting it from the main solid fails with BRep_NotDone.
         # the work around is to extrude each and then join the resulting solids, which seems to work
 
         # underlying cad kernel can only handle simple bosses-- we'll aggregate them if there are
         # multiple sets
+        thisObj: Union[Solid, Compound]
 
         toFuse = []
+        taper = 0.0 if taper is None else taper
+        baseSolid = None
 
-        if taper:
-            for ws in wireSets:
-                thisObj = Solid.extrudeLinear(ws[0], [], eDir, taper)
+        for ws in wireSets:
+            if upToFace is not None:
+                baseSolid = self.findSolid() if baseSolid is None else thisObj
+                if isinstance(upToFace, int):
+                    facesList = getFacesList(eDir, additive, both=both)
+                    limitFace = facesList[upToFace]
+                    if (
+                        baseSolid.isInside(ws[0].Center())
+                        and additive
+                        and upToFace == 0
+                    ):
+                        upToFace = 1  # extrude until next face outside the solid
```

Comment:
I may be misunderstanding your intention here, but I think you meant:
```suggestion
                    if (
                        baseSolid.isInside(ws[0].Center())
                        and additive
                        and upToFace == 0
                    ):
                        upToFace = 1  # extrude until next face outside the solid
                        
                    limitFace = facesList[upToFace]
```

I would suggest a test is added for this as well.

Reply:
I think you misunderstood, this part of the code does : 
- A check if `upToface` is of type `int` or `Face` , if it's type `int` it checks that if an extrusion is required and the extrusion until the next face is required, if the wire is insinde the solid then we want to extrude until the next face **Outside** the solid. Hence the `upToFace = 1`
- If one of these conditions isn't met then we are fine with `upToFace = 0`
- And finally if ``upToFace isn't of type `int` then  it's of type `Face` and we pass the face to extrude to directly

There is indeed no check if `upToFace` is actually of type `Union[int, Face]`. Would you want to add a type check at the beginning and raise an error if that is not the case ? (And add a test associated with it)

I don't know if it's relevant since the allowed type are given in the method signature and it's a private method. A type check is already performed on the public method `extrude`

Reply:
I didn't mean types. 

> if the wire is insinde the solid then we want to extrude until the next face Outside the solid. Hence the upToFace = 1

Then you probably want to set `upToFace = 1` *before* you do `limitFace = facesList[upToFace]`?